### PR TITLE
lecture-5-checkoutpage

### PR DESCRIPTION
**Pull request by IvanZenov**

### CODE REVIEW

File: src/main/java/com/es/phoneshop/dao/impl/OrderDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        }
+        return INSTANCE;
+    }
+
+    @Override
+    public Order getOrder(String secureId) throws ProductNotFoundException {
+        return super.entities.stream()
+                .filter(order -> order.getSecureId().equals(secureId))
+                .findAny()
+                .orElseThrow(() -> new ProductNotFoundException("Order not found"));
```
Comment (nd524s) [inline src/main/java/com/es/phoneshop/dao/impl/OrderDaoImpl.java]: Throw OrderNotFoundException here

File: src/main/java/com/es/phoneshop/model/entity/BaseEntity.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package com.es.phoneshop.model.entity;
+
+import java.io.Serializable;
+
+public class BaseEntity implements Serializable {
```
Comment (nd524s) [inline src/main/java/com/es/phoneshop/model/entity/BaseEntity.java]: Maybe you can do something like this in class declaration:
public class BaseEntity implements Serializable, Cloneable

You can avoid repetition of  "implements Serializable, Cloneable" in children classes

File: src/main/java/com/es/phoneshop/service/impl/OrderServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    @Override
+    public Order getOrder(Cart cart) {
+        Order order = new Order();
+
+        order.setCartItemList(cart.getItems().stream().map(item -> {
+            try {
+                return (CartItem) item.clone();
+            } catch (CloneNotSupportedException ex) {
+                throw new RuntimeException();
```
Comment (nd524s) [inline src/main/java/com/es/phoneshop/service/impl/OrderServiceImpl.java]: Throw more specific RuntimeException with clear message what happened.

File: src/main/java/com/es/phoneshop/security/DosServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package com.es.phoneshop.security;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class DosServiceImpl implements DefaultDosService {
+
+    private static final long THRESHOLD = 10;
```
Comment (nd524s) [inline src/main/java/com/es/phoneshop/security/DosServiceImpl.java:9]: I think you need to add here a time condition as well. It's required by task

**Review by nd524s — CHANGES_REQUESTED** [review]

File: src/main/java/com/es/phoneshop/model/entity/BaseEntity.java (hunk 2 above)
Comment (IvanZenov) [reply src/main/java/com/es/phoneshop/model/entity/BaseEntity.java]: fix

File: src/main/java/com/es/phoneshop/security/DosServiceImpl.java (hunk 4 above)
Comment (IvanZenov) [reply src/main/java/com/es/phoneshop/security/DosServiceImpl.java:9]: fix

File: src/main/java/com/es/phoneshop/dao/impl/OrderDaoImpl.java (hunk 1 above)
Comment (IvanZenov) [reply src/main/java/com/es/phoneshop/dao/impl/OrderDaoImpl.java]: fix

File: src/main/java/com/es/phoneshop/service/impl/OrderServiceImpl.java (hunk 3 above)
Comment (IvanZenov) [reply src/main/java/com/es/phoneshop/service/impl/OrderServiceImpl.java]: fix

**Review by nd524s — APPROVED** [review]

**Pull request merged**